# 4.6 TASK - Combining and Exporting Data

1. Import data sets
2. check dimensions
3. determine suitable way to combine orders_products_combined df with products data
4. confirm results with merge flag
5. export in sutable format

In [1]:
import pandas as pd
import numpy as np 
import os

1 import data sets into jupyter

In [2]:
# forge main path
path = r"C:\Users\Noah\4.3.24_INSTACART BASKET ANALYSIS"

In [3]:
path

'C:\\Users\\Noah\\4.3.24_INSTACART BASKET ANALYSIS'

In [4]:
# import data set ordsprodscombined pickle
df_ords_prods_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

In [5]:
# import data set products_checked.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

2. check dimensions

In [6]:
df_ords_prods_combined.head()

order_id  user_id eval_set  order_number  orders_day_of_the_week  \
0   2539329        1    prior             1                       2   
1   2539329        1    prior             1                       2   
2   2539329        1    prior             1                       2   
3   2539329        1    prior             1                       2   
4   2539329        1    prior             1                       2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [7]:
#check dimensions to ensure import matches export
df_ords_prods_combined.shape

(32434489, 11)

Drop Merge column from df_ords_prods_combined so it can be merged with df_prods

In [8]:
df_ords_prods_combined = df_ords_prods_combined.drop(['_merge'], axis = 1)

In [9]:
df_ords_prods_combined.head()

order_id  user_id eval_set  order_number  orders_day_of_the_week  \
0   2539329        1    prior             1                       2   
1   2539329        1    prior             1                       2   
2   2539329        1    prior             1                       2   
3   2539329        1    prior             1                       2   
4   2539329        1    prior             1                       2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   

   reordered  
0          0  
1          0  
2          0  
3          0  
4          0

In [10]:
df_prods.shape

(49670, 6)

In [11]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [12]:
#drop unnamed column
df_prods = df_prods.drop(['Unnamed: 0'], axis = 1)

In [13]:
#recheck
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [14]:
# ensuring shape matches minus one column
df_prods.shape

(49670, 5)

 3. determine suitable method for combining df_ords_prods_combined and df_prods

 There is a shared "product_id" column onto which the data can be merged. Using an inner join here

In [15]:
df_ords_prods_merge = df_prods.merge(df_ords_prods_combined, on = 'product_id', indicator = True)

In [16]:
#check head of conbined data
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_the_week  \
0   3139998      138    prior            28                       6   
1   1977647      138    prior            30                       6   
2    389851      709    prior             2                       0   
3    652770      764    prior             1                       3   
4   1813452      764    prior             3                       4   

   order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                 11                     3.0                  5          0   
1                 17                    20.0                  1          1   
2                 21                     6.0                 20          0   
3                 13                     NaN                 10          0   
4                 17                     9.0                 11          1   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

In [17]:
# checking shape of combined data
df_ords_prods_merge.shape

(32399732, 15)

4. confirm results with merge flag

In [18]:
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32399732
left_only            0
right_only           0
Name: count, dtype: int64

# the final output has 32,399,732 rows. Each row has info found in both the input data sets

5. export data

In [19]:
# export as pickle
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', '4.6_ords_prods_merge.pkl'))

In [20]:
df_ords_prods_merge.head(0)

Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices, order_id, user_id, eval_set, order_number, orders_day_of_the_week, order_hour_of_day, days_since_prior_order, add_to_cart_order, reordered, _merge]
Index: []

In [24]:
print(df_ords_prods_merge.head(0))

Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices, order_id, user_id, eval_set, order_number, orders_day_of_the_week, order_hour_of_day, days_since_prior_order, add_to_cart_order, reordered, _merge]
Index: []
